### PR TITLE
LibC: Fixing `assert`

### PR DESCRIPTION
**Pull request by mhermier**
Hi,
The following change set fix some issues in `LibC` `assert`:
- declare `_abort` as a C linkage
- Make private API always available even if `LibC` was compiled with `NDEBUG` set.
- remove `__stringify` macro by using a `__assert` private API more in part with what glibc provides (glibc headers mention that this is required for "standard compliance" but I was not able to find the standard...).

### CODE REVIEW

**Review by ADKaster — CHANGES_REQUESTED** [review]

File: Userland/Libraries/LibC/assert.cpp [hunk 1]
```diff
@@ -16,13 +16,13 @@
 extern "C" {
 
 extern bool __stdio_is_initialized;
-#ifndef NDEBUG
```
[the remark this answers is not in the document]
Comment (BertalanD) [reply Userland/Libraries/LibC/assert.cpp:19]: > Do we need to add an __attribute__((used)) to this guy to make sure it's always exported/included in the library?

I don't think so. We compile with default symbol visibility, so all non-`static` functions are exported from shared libraries implicitly.

File: Userland/Libraries/LibC/assert.cpp [hunk 2]
```diff
@@ -16,13 +16,13 @@
...
-void __assertion_failed(const char* msg)
+
+void __assert(const char* assertion, const char* file, int line)
```
[the remark this answers is not in the document]
Comment (mhermier) [reply Userland/Libraries/LibC/assert.cpp:20]: This is the obvious place I looked first, so maybe it is an informal convention between implementations, or some historical unix stuff...

File: Userland/Libraries/LibC/stdlib.h [hunk 3]
```diff
@@ -10,14 +10,14 @@
...
-__attribute__((warn_unused_result)) int __generate_unique_filename(char* pattern);
-
 __BEGIN_DECLS
 
 #define EXIT_SUCCESS 0
 #define EXIT_FAILURE 1
 #define MB_CUR_MAX 4
 
+__attribute__((warn_unused_result)) int __generate_unique_filename(char* pattern);
+
```
Comment (BertalanD) [inline Userland/Libraries/LibC/stdlib.h:20]: Why is this change needed? This function isn't even defined anywhere, as it's been replaced by the un-uglified `generate_unique_filename` template, which is only used in `stdlib.cpp`. A better change would be to remove this declaration completely.

**Comment by BertalanD** [comment]
This PR is titled "Fixing `assert`. Are you still planning to implement the changes you and Andrew discussed in the comments?

**Comment by mhermier** [comment]
Yes, the thing is that I have trouble with my environment, and can only rely on the qt creator internal compiler/linter to code. So I have to push the change set gradually to prove each commits with CI...

File: Userland/Libraries/LibC/stdlib.h (hunk 3 above)
Comment (mhermier) [reply Userland/Libraries/LibC/stdlib.h:20]: Not related, I can make a separated PR about this. I found it when I needed to move `_fork` to stdlib, in a patch not pushed yet.

File: Userland/Libraries/LibC/assert.cpp [hunk 4]
```diff
@@ -13,8 +13,6 @@
 #include <syscall.h>
 #include <unistd.h>
 
-extern "C" {
-
```
Comment (BertalanD) [inline Userland/Libraries/LibC/assert.cpp:17]: I don't follow the rationale of this commit.  We tend to put both the definitions and declarations of functions that might be called from C code inside `extern "C"` blocks.

Comment (mhermier) [reply Userland/Libraries/LibC/assert.cpp:17]: I'm OK, I'll change that then.

**Comment by BertalanD** [comment]
Also, please note that this does not compile. I'd gladly help you figure out any build issues you might have if you reach me on Discord. My name is `BertalanD` there too.

**Comment by mhermier** [comment]
Sorry, but I'm not a discord user. I'm an old IRC dude XD to be more real, I limit my social network on purpose (my year wish of 2 years ago)

**Comment by awesomekling** [comment]
> Yes, the thing is that I have trouble with my environment, and can only rely on the qt creator internal compiler/linter to code. So I have to push the change set gradually to prove each commits with CI...

Please don't push code to a non-draft PR if you haven't even built it locally. It slows down CI for everybody else and spams people with notifications.

**Comment by mhermier** [comment]
Well, if you could fix "Meta/serenity.sh x86_64 Clang" to not rely on gcc 10.x to build host tools, I could do local tests. Unless code rely on C++20 stdlib header (which I think it is not), I think it should be done... (I was able to do compile build before by bring gcc from a developer repos on my distro, but I had to revert back because of instabilities...)

**Comment by ADKaster** [comment]
The host tools require a C++20 enabled gcc or clang host compiler. You can use gcc 10.2 or higher, or clang 13 or higher. They use C++20 language features because the AK and LibCore libraries use C++20 features. You can also tailor the host build to use whichever compiler you like, via running the commands manually per the AdvancedBuildInstructions section on the super build. But again, like Bertalan mentioned, the best way to get your build figured out is over discord.

**Comment by mhermier** [comment]
Finally took some time to deal with that issue and it seems I found a way to make a crude hack to `Meta/serenity.sh` to replace gcc with clang. I tried it before without success, and was not considering that option as valid. Will see if that is enough for a full build, and perform proper tests.

**Comment by mhermier** [comment]
I renamed the branch and it automatically closed the pull request which is not what I expected... I'll do a new pull request on this when I get things sorted out. In the mean time, I'll do pull request on the more trivial things I found investigating this.
